File: JAVA-Workshop/war/src/main/java/fr/epita/assistants/war/Soldier.java
package fr.epita.assistants.war;

public class Soldier extends Combatant{
    protected int health;
    protected int damage;
    protected String scream;

    public Soldier() {
        this.health = 15;
        this.damage = 3;
        this.scream = "No pity for losers!";
    }

    public void kill() {
        this.health = 0;
    }

    @Override
    public void printState() {
        System.out.println("I have " + this.health + " health points.");
    }

    @Override
    public void attack(Soldier s) {
       s.health -= this.damage;
    }

    @Override
    void attack(Vehicle v) {
        System.out.println("I can't fight this.");
    }

    @Override
    public void scream() {
        System.out.println(this.scream);
    }
}

File: CPP-Workshop/war/soldier.cc
#include "soldier.hh"

#include <iostream>

Soldier::Soldier()
    : hp(15)
    , damage(3)
{}

Soldier::Soldier(int hp, int damage)
    : hp(hp)
    , damage(damage)
{}

void Soldier::attack(Soldier& s)
{
    s.hp -= this->damage;
}

void Soldier::print_state() const
{
    std::cout << "I have " << this->hp << " health points.\n";
}

void Soldier::scream() const
{
    std::cout << "No pity for losers!\n";
}
